pilot/pkg/networking/plugin/envoyfilter/envoyfilter.go

```diff
@@ -0,0 +1,301 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package envoyfilter
+
+import (
+	"net"
+	"strings"
+
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2/listener"
+	http_conn "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/http_connection_manager/v2"
+
+	networking "istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/plugin"
+	"istio.io/istio/pkg/log"
+)
+
+type envoyfilterplugin struct{}
+
+const (
+	// InboundDirection indicates onInboundListener callback
+	InboundDirection = "inbound"
+	// OutboundDirection indicates onOutboundListener callback
+	OutboundDirection = "outbound"
+)
+
+// NewPlugin returns an ptr to an initialized envoyfilter.Plugin.
+func NewPlugin() plugin.Plugin {
+	return envoyfilterplugin{}
+}
+
+func getListenerIPAddress(address *core.Address) net.IP {
+	if address != nil && address.Address != nil {
+		switch t := address.Address.(type) {
+		case *core.Address_SocketAddress:
+			if t.SocketAddress != nil {
+				ip := "0.0.0.0"
+				if t.SocketAddress.Address != "::" {
+					ip = t.SocketAddress.Address
+				}
+				return net.ParseIP(ip)
+			}
+		}
+	}
+	return nil
+}
+
+func listenerMatch(in *plugin.InputParams, direction string, listenerIP net.IP,
```

Comment:
code in this file would really benefit from unit tests. there's lots of stuff going on that could be error prone and would be relatively easy to test.